File: toy/e-commerce-java/src/main/java/me/kzv/ecommerce/entity/Item.java
package me.kzv.ecommerce.entity;

import javax.persistence.*;
import lombok.Getter;
import lombok.Setter;


import lombok.ToString;
import me.kzv.ecommerce.constant.ItemSellStatus;
import me.kzv.ecommerce.dto.ItemFormDto;
import me.kzv.ecommerce.exception.OutOfStockException;


@Entity
@Table(name="item")
@Getter
@Setter
@ToString
public class Item extends BaseEntity {

    @Id
    @Column(name="item_id")
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id; //상품 코드

    @Column(nullable = false, length = 50)
    private String itemNm; //상품명

    @Column(name="price", nullable = false)
    private int price; //가격

    @Column(nullable = false)
    private int stockNumber; //재고수량

    @Lob
    @Column(nullable = false)
    private String itemDetail; //상품 상세 설명

    @Enumerated(EnumType.STRING)
    private ItemSellStatus itemSellStatus; //상품 판매 상태

    public void updateItem(ItemFormDto dto) {
        this.itemNm = dto.getItemNm();
        this.price = dto.getPrice();
        this.stockNumber = dto.getStockNumber();
        this.itemDetail = dto.getItemDetail();
        this.itemSellStatus = dto.getItemSellStatus();
    }

    public void removeStock(int stockNumber) {
        int restStock = this.stockNumber - stockNumber;
        if (restStock < 0) {
            throw new OutOfStockException("상품의 재고가 부족 합니다.");
        }

        this.stockNumber = restStock;
    }

    public void addStock(int stockNumber) {
        this.stockNumber += stockNumber;
    }
}

File: toy/e-commerce/src/main/kotlin/me/kzv/shop/entity/Item.kt
package me.kzv.shop.entity

import me.kzv.shop.entity.constant.ItemSellStatus
import me.kzv.shop.exception.OutOfStockException
import javax.persistence.*

@Entity
class Item(

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "item_id")
    var id: Long? = null,

    var itemName: String, // 상품명

    var price: Int, // 상품 가격
    // 포인트나 할인제도가 추가 될경우 ...

    var stockNumber: Int, // 재고 수량

    @Lob
    var itemDetail: String, // 상품 상세 설명

    @Enumerated(EnumType.STRING)
    var itemSellStatus: ItemSellStatus, //판매 상태

) : BaseEntity() {
    fun updateItem() {

    }

    fun removeStock(stockNumber: Int) {
        val restStock = this.stockNumber - stockNumber
        if (restStock < 0) {
            throw OutOfStockException("상품의 재고가 부족 합니다. (현재 재고 수량: ${this.stockNumber})")
        }

        this.stockNumber = restStock
    }

    fun addStock(stockNumber: Int){
        this.stockNumber = stockNumber
    }
}